### PR TITLE
feat(share/shrex): add Rate limit middleware for shrex servers

### PR DESCRIPTION
**Pull request by walldiss**
## Overview
Closes https://github.com/celestiaorg/celestia-node/issues/1747

Add concurrency rate limiting on server side of shrex. 

## Checklist

- [ ] New and updated code has appropriate documentation
- [ ] New and updated code has new and/or updated testing
- [ ] Required CI checks are passing
- [ ] Visual proof for any user facing features like CLI or documentation updates
- [ ] Linked issues closed with keywords

### CODE REVIEW

**Comment by distractedm1nd** [comment]
Hey, I will merge this into the integration branch and run some tests with it before I approve, but looks really good

File: share/p2p/middleware.go
Comment (Wondertan) [inline share/p2p/middleware.go:27]: I think we had a special error that was supposed to be sent in this case and not just stream closage

Comment (distractedm1nd) [reply share/p2p/middleware.go:27]: We did, but we want to rethink this design - this would still require the overloaded nodes to read the request from the client, and then send a response - it is much better if we can just close the stream gracefully, and have the client interpret it as a rate limit response

**Comment by distractedm1nd** [comment]
Works beautifully on TG

File: share/p2p/middleware.go
Comment (distractedm1nd) [reply share/p2p/middleware.go:27]: @Wondertan Can we merge this for now? We won't merge it until we have your okay, but we do want it in the feature branch. It significantly improves metrics

Comment (walldiss) [reply share/p2p/middleware.go:27]: As far as I remember initially @Wondertan purposed to Close the stream without response, but we revised it later with idea of sending response with status instead. Motivation was that it would allow us to know if stream was unintentionally closed or it was expected rate limiting. 

With closer look it turned out we can identify graceful stream close with error handling check, so there is no need to send extra message.

Pros of getting rid of extra message:
With explicit message approach we would have an extra round trip. Server that is already exhausted, would need to read request first, which will be thrown away. And then it would need to send response containing "Status_REFUSED". We could achieve the same, without extra round trip.

Also we can have rate limiting protocol agnostic, as it would not need to read out protocol specific request.

Comment (Wondertan) [reply share/p2p/middleware.go:27]: The reasoning makes sense to me. Thanks for explaining. As long as there is explicit handling for the client side of this case I am ok with this approach. We will also need to update the specs in the Adr right after.

**Review by Wondertan — APPROVED** [review]

File: share/p2p/shrexeds/server.go
Comment (distractedm1nd) [inline share/p2p/shrexeds/server.go:67]: we can remove this right

**Review by distractedm1nd — CHANGES_REQUESTED** [review]

File: share/p2p/shrexeds/server.go
Comment (walldiss) [reply share/p2p/shrexeds/server.go]: yep

**Review by distractedm1nd — APPROVED** [review]

File: share/p2p/shrexnd/options.go
Comment (renaynay) [inline share/p2p/shrexnd/options.go:60]: why not errSuffix

**Review by renaynay — APPROVED** [review]

**Pull request merged**
